lsp-mode.el

```diff
@@ -2028,6 +2030,223 @@ CALLBACK - callback for the lenses."
     (setq-local lsp--lens-last-count nil)
     (setq-local lsp--lens-backend-cache nil))))
 
+
+;; toggles
+
+(defun lsp-toggle-trace-io ()
+  "Toggle client-server protocol logging."
+  (interactive)
+  (setq lsp-print-io (not lsp-print-io))
+  (lsp--info "Server logging %s." (if lsp-print-io "enabled" "disabled")))
+
+(defun lsp-toggle-signature-auto-activate ()
+  "Toggle signature auto activate."
+  (interactive)
+  (setq lsp-signature-auto-activate (not lsp-signature-auto-activate))
+  (lsp--info "Signature autoactivate %s." (if lsp-signature-auto-activate "enabled" "disabled"))
+  (lsp--update-signature-help-hook))
+
+(defun lsp-toggle-on-type-formatting ()
+  "Toggle on type formatting."
+  (interactive)
+  (setq lsp-enable-on-type-formatting (not lsp-enable-on-type-formatting))
+  (lsp--info "On type formatting is %s." (if lsp-enable-on-type-formatting "enabled" "disabled"))
+  (lsp--update-on-type-formatting-hook))
+
+(defun lsp-toggle-symbol-highlight ()
+  "Toggle symbol highlighting."
+  (interactive)
+  (setq lsp-enable-symbol-highlighting (not lsp-enable-symbol-highlighting))
+
+  (cond
+   ((and lsp-enable-symbol-highlighting
+         (lsp-feature? "textDocument/documentHighlight"))
+    (add-hook 'lsp-on-idle-hook #'lsp--document-highlight nil t)
+    (lsp--info "Symbol highlighting enabled in current buffer."))
+   ((not lsp-enable-symbol-highlighting)
+    (remove-hook 'lsp-on-idle-hook #'lsp--document-highlight t)
+    (lsp--remove-overlays 'lsp-highlight)
+    (lsp--info "Symbol highlighting disabled in current buffer."))))
+
+
+;; keybindings
+
+(defmacro lsp-define-conditional-key (keymap key def cond &rest bindings)
+  "In KEYMAP, define key sequence KEY as DEF conditionally.
+This is like `define-key', except the definition disappears
+whenever COND evaluates to nil.
+BINDINGS is a list of (key def cond)."
+  (declare (indent defun)
+           (debug (form form form form &rest sexp)))
+  (->> (cl-list* key def cond bindings)
+       (-partition 3)
+       (-map (-lambda ((key def cond))
+               `(define-key ,keymap ,key
+                  '(menu-item
+                    ,(format "maybe-%s" def)
+                    ,def
+                    :filter (lambda (item)
+                              (when (with-current-buffer lsp--describe-buffer
+                                      ,cond)
+                                item))))))
+       macroexp-progn))
```

Comment:
This looks nice. Do you need `(require 'macroexp)` too?

Reply:
Probably something that we require already loads this package or the byte-compilation tests would fail. I guess it won't be an issue if we add it.